File: platforms/android/stato/src/main/java/com/facebook/stato/plugins/leakcanary/LeakCanaryStatoPlugin.java
/*
 *  Copyright (c) 2018-present, Facebook, Inc.
 *
 *  This source code is licensed under the MIT license found in the LICENSE
 *  file in the root directory of this source tree.
 *
 */

package com.facebook.stato.plugins.leakcanary;

import android.util.Log;
import com.facebook.stato.core.StatoConnection;
import com.facebook.stato.core.StatoObject;
import com.facebook.stato.core.StatoPlugin;
import com.facebook.stato.core.StatoReceiver;
import com.facebook.stato.core.StatoResponder;
import java.util.ArrayList;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class LeakCanaryStatoPlugin implements StatoPlugin {

  private static final String TAG = "LeakCanaryStatoPlugin";

  private static final String REPORT_LEAK_EVENT = "reportLeak";
  private static final String CLEAR_EVENT = "clear";
  private static final String LEAKS_KEY = "leaks";

  private StatoConnection mConnection;

  private final List<String> leakList = new ArrayList<>();

  @Override
  public String getId() {
    return "@stato/plugin-leak-canary";
  }

  @Override
  public void onConnect(StatoConnection connection) {
    mConnection = connection;
    sendLeakList();

    mConnection.receive(
        CLEAR_EVENT,
        new StatoReceiver() {
          @Override
          public void onReceive(StatoObject params, StatoResponder responder) throws Exception {
            leakList.clear();
          }
        });
  }

  @Override
  public void onDisconnect() throws Exception {
    mConnection = null;
  }

  @Override
  public boolean runInBackground() {
    return false;
  }

  private void sendLeakList() {
    if (mConnection != null) {
      JSONObject obj = new JSONObject();
      try {
        obj.put(LEAKS_KEY, new JSONArray(leakList));
        mConnection.send(REPORT_LEAK_EVENT, new StatoObject(obj));
      } catch (JSONException e) {
        Log.w(TAG, "Failure to serialize leak list: ", e);
      }
    }
  }

  public void reportLeak(String leakInfo) {
    leakList.add(leakInfo);
    sendLeakList();

  }
}

File: platforms/android/android-stato/src/main/kotlin/org/stato/plugins/leakcanary/LeakCanaryStatoPlugin.kt
/*
 *  Copyright (c) 2018-present, Facebook, Inc.
 *
 *  This source code is licensed under the MIT license found in the LICENSE
 *  file in the root directory of this source tree.
 *
 */

package org.stato.plugins.leakcanary

import android.util.Log
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.stato.core.*

class LeakCanaryStatoPlugin : StatoPlugin {



  class Factory : StatoPluginFactory<LeakCanaryStatoPlugin> {
    override val id = ID

    override fun create() =
      LeakCanaryStatoPlugin()
  }

  private var connection: StatoPluginConnection? = null

  private val leakList = mutableListOf<String>()

  override val id = ID

    override fun onConnect(connection: StatoPluginConnection) {
    this.connection = connection
    sendLeakList()

    this.connection!!.receive(CLEAR_EVENT) { params: StatoObject, responder: StatoResponder ->
          leakList.clear()
        }
      }


  @Throws(Exception::class)
  override fun onDisconnect() {
    connection = null
  }

    override fun runInBackground(): Boolean {
    return false
  }

  private fun sendLeakList() {
    connection?.run {

      try {

        send(REPORT_LEAK_EVENT, StatoObject(JSONObject().put(LEAKS_KEY, JSONArray(leakList))))
      } catch (e: JSONException) {
        Log.w(TAG, "Failure to serialize leak list: ", e)
      }
    }

  }

  fun reportLeak(leakInfo: String) {
    leakList.add(leakInfo)
    sendLeakList()

  }

  companion object {
    const val ID = "@stato/plugin-leak-canary"
    private val TAG = "LeakCanaryStatoPlugin"

    private val REPORT_LEAK_EVENT = "reportLeak"
    private val CLEAR_EVENT = "clear"
    private val LEAKS_KEY = "leaks"
  }
}
